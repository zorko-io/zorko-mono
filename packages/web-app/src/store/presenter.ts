export interface Presenter<S, I> {
    runMutations(): this
    toJS(): S
    toImmutable(): I
}


export interface Getter {}

