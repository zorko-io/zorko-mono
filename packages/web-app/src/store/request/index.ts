export * from './request.presenter'
export * from './request.state';
