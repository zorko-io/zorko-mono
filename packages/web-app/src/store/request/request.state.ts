export interface RequestState {
  isPending: boolean
  isSucceed: boolean
  error: Error | null
}
