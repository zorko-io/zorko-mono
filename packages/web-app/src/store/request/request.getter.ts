import { Getter } from '../presenter';
import { Map } from 'immutable';

export class RequestGetter implements Getter {
  static create(){
    return new RequestGetter();
  }

  readonly isPendingPath: string;
  readonly isSucceedPath: string;
  readonly errorPath: string;

  constructor() {
    this.isPendingPath = 'isPending';
    this.isSucceedPath = 'isSucceed';
    this.errorPath = 'error';
  }

  isPending = (state: Map<string, any>) =>  {
    return state.get(this.isPendingPath);
  };

  isSucceed = (state: Map<string, any>) => {
    return state.get(this.isSucceedPath);
  };

  getError = (state: Map<string, any>) => {
    return state.get(this.errorPath);
  };
}
