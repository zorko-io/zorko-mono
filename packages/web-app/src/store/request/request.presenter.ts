import _ from 'lodash';
import { Map, fromJS } from 'immutable';
import { Presenter } from '../presenter';

export interface RequestState {
  isPending: boolean
  isSucceed: boolean
  error: Error | null
}

const DEFAULT_REQUEST_STATE: RequestState = {
  isPending: false,
  isSucceed: false,
  error: null
};

export class RequestStateImmutable implements Presenter<RequestState> {
  private state: Map<string, any>;
  private mutations: Function[];

  static fromJS(state: Partial<RequestState>): RequestStateImmutable{
    return RequestStateImmutable.create(fromJS({
      ...DEFAULT_REQUEST_STATE,
      ...state,
    }));
  }

  static create(state?: Map<string, any>): RequestStateImmutable {
    return new RequestStateImmutable(state || fromJS(DEFAULT_REQUEST_STATE));
  }

  constructor(state: Map<string, any>) {
    this.state = state;
    this.mutations = [];
  }

  isPending(): boolean {
    return this.state.get('isPending');
  }

  isSucceed(): boolean {
    return this.state.get('isSucceed');
  }

  getError (): Error | null {
    return this.state.get('error');
  }

  setPending(value: boolean) {
    this.mutations.push((state: Map<string, any>) => state.set('isPending', value));
    return this;
  }

  setSucceed(value: boolean) {
    this.mutations.push(
      (state: Map<string, any>) => state
        .set('isPending', false)
        .set('error', null)
        .set('isSucceed', value)
    );
    return this;
  }

  runMutations () {
    this.state = this.state.withMutations(state => {
        this.mutations.forEach(mutation => {
          mutation(state);
        })
    });
    this.mutations = [];
    return this;
  }

  setError(value: Error | null) {
    this.mutations.push(
      (state: Map<string, any>) => state
        .set('isPending', false)
        .set('isSucceed', false)
        .set('error', fromJS(value))
    );

    return this;
  }

  reset() {
    this.state = fromJS(DEFAULT_REQUEST_STATE);
    this.mutations = [];

    return this;
  }

  toImmutable(): Map<string, any> {
    if (!_.isEmpty(this.mutations)){
      this.runMutations();
    }

    return this.state;
  }

  toJS(): RequestState {
    if (!_.isEmpty(this.mutations)){
      this.runMutations();
    }

    return {
      isPending: this.isPending(),
      isSucceed: this.isSucceed(),
      error: this.getError()
    }
  }

}

export class RequestPresenter {

  private immutable: Map<string, any>;

  static fromJS(state?: RequestState) {
    return new RequestPresenter(state ? fromJS(state) : state)
  }

  static create(state?: Map<string, any>) {
    return new RequestPresenter(state)
  }

  constructor(state?: Map<string, any>){
    if (!state){
      state = Map({
        isPending: false,
        isSucceed: false,
        error: null
      })
    }

    /// TODO: validation !!???

    this.immutable = state;
  }

  startRequest(){
    this.immutable = this.immutable.set('isPending', true);
    return this;
  }

  markSucceed () {
    this.immutable = this.immutable
        .set('isPending', false)
        .set('isSucceed', true);

    return this;
  }

  markFailure(error: Error) {
    this.immutable = this.immutable
      .set('isPending', false)
      .set('isSucceed', false)
      .set('error', fromJS(error));

    return this;
  }

  reset() {
    this.immutable = this.immutable
      .set('isPending', false)
      .set('isSucceed', false)
      .set('error', null);

    return this;
  }

  toJS(): RequestState {
    return this.immutable.toJS() as RequestState;
  }

  toImmutable(){
    return this.immutable;
  }


}
