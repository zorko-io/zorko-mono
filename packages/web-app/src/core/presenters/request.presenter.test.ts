import { RequestPresenter } from './request.presenter';

describe('RequestPresenter', () => {

  let presenter: RequestPresenter;

  beforeEach(()=> {
     presenter = RequestPresenter.create();
  });

  it('creates presenter with defaults', () => {
    let expected = presenter.toJS();
    expect(expected).toMatchSnapshot();
    expect(presenter.toImmutable().toJS()).toEqual(expected);
  });

  it('starts request', () => {
    let actual = presenter.startRequest();

    expect(presenter).toBe(actual);

    let expected = actual.toJS();
    expect(expected).toMatchSnapshot();
    expect(actual.toImmutable().toJS()).toEqual(expected);
  });

  it('marks as succeed request', () => {
    let actual = presenter.startRequest().markSucceed();

    expect(presenter).toBe(actual);
    let expected = actual.toJS();
    expect(expected).toMatchSnapshot();
    expect(actual.toImmutable().toJS()).toEqual(expected);
  }) ;

  it('marks as error request', () => {
    let actual = presenter
      .startRequest()
      .markFailure(new Error('Boom!'));

    expect(presenter).toBe(actual);
    let expected = actual.toJS();
    expect(expected).toMatchSnapshot();
    expect(actual.toImmutable().toJS()).toEqual(expected);
  });

  it('resets to default', () => {
    let actual = presenter
      .startRequest()
      .markFailure(new Error('Boom!'))
      .reset();

    expect(presenter).toBe(actual);
    let expected = actual.toJS();
    expect(expected).toMatchSnapshot();
    expect(actual.toImmutable().toJS()).toEqual(expected);
  })

});
