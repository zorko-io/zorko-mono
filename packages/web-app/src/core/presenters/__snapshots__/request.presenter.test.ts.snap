// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestPresenter creates presenter with defaults 1`] = `
Object {
  "error": null,
  "isPending": false,
  "isSucceed": false,
}
`;

exports[`RequestPresenter marks as error request 1`] = `
Object {
  "error": [Error: Boom!],
  "isPending": false,
  "isSucceed": false,
}
`;

exports[`RequestPresenter marks as succeed request 1`] = `
Object {
  "error": null,
  "isPending": false,
  "isSucceed": true,
}
`;

exports[`RequestPresenter resets to default 1`] = `
Object {
  "error": null,
  "isPending": false,
  "isSucceed": false,
}
`;

exports[`RequestPresenter starts request 1`] = `
Object {
  "error": null,
  "isPending": true,
  "isSucceed": false,
}
`;
